src/mlpack/methods/lars/lars_impl.hpp

```diff
@@ -375,20 +520,32 @@ inline double LARS::Train(const arma::mat& matX,
         const double dirCorr = dot(dataRef.col(ind), yHatDirection);
         const double val1 = (maxCorr - corr(ind)) / (normalization - dirCorr);
         const double val2 = (maxCorr + corr(ind)) / (normalization + dirCorr);
-        if ((val1 > 0.0) && (val1 < gamma))
-          gamma = val1;
-        if ((val2 > 0.0) && (val2 < gamma))
-          gamma = val2;
-        // Handle edge case where the largest actually is equal to 0.
-        if (std::max(val1, val2) == 0.0)
-          gamma = 0.0;
+
+        // If we kicked out a feature due to the LASSO modification last
+        // iteration, then we do not allow relaxation of the step size to 0 for
+        // that feature in this iteration.
+        if (lassocond && (ind == lassocondInd))
+        {
+          if ((val1 > 0.0) && (val1 < gamma))
+            gamma = val1;
+          if ((val2 > 0.0) && (val2 < gamma))
+            gamma = val2;
+        }
+        else
+        {
+          if ((val1 >= 0.0) && (val1 < gamma))
+            gamma = val1;
+          if ((val2 >= 0.0) && (val2 < gamma))
+            gamma = val2;
+        }
```

Comment:
It turns out the "tiny relaxation" is not so tiny: on LARS iterations where we are using the LASSO modification, sometimes we remove a feature because its sign changes (this is Eq. 3.6).  However, we need to avoid considering it for the next iteration---so we can't use the relaxation that allows `gamma` (the step size) to become 0, since the removed feature's correlation should be exactly zero.